helix-vcs/src/git.rs

```diff
@@ -51,6 +104,111 @@ impl Git {
         )
         .ok()
     }
+
+    /// Emulates the result of running `git status` from the command line.
+    fn status(repo: &Repository) -> Result<Vec<FileChange>> {
+        let autocrlf = repo
+            .config_snapshot()
+            .boolean("core.autocrlf")
+            .unwrap_or(false);
+
+        let work_dir = repo
+            .work_dir()
+            .ok_or_else(|| anyhow::anyhow!("working tree not found"))?;
+
+        // TODO: allow diffing against another ref
+        let head_tree = repo.head_commit()?.tree()?;
+
+        let mut head_tree_files = vec![];
+        let mut submodule_paths = vec![PathBuf::from(".git")];
+        traverse_tree(
```

Comment:
Instead of traversing the tree, using [the index](https://docs.rs/git-repository/0.33.0/git_repository/struct.Repository.html#method.index) is the better and faster way to going about these things - it can be assumed to be up-to-date as git updates it to match any checked out commit.

If the index was used, the entire tree-traversal part could be replaced with a loop over index entries which should remove some complexity.

If the `allow diffing against another ref` feature would have to be implemented, tree traversal would be a must - in that case I recommend creating [an index from a tree](https://docs.rs/git-index/0.12.2/git_index/struct.State.html#method.from_tree) so the main code-path deals with [the index](https://docs.rs/git-index/0.12.2/git_index/struct.State.html#) as primary data structure only.

Reply:
Initially I wanted to implement that too bit aren't you then diffing between the index/stage instead of the work tree (so git status --cached)?

Reply:
> aren't you then diffing between the index/stage instead of the work tree

Unless I can create the index from the work tree? I'll look into it a bit. Thanks @Byron for the pointer!

Reply:
That's true, sorry for that, maybe that's not even intended. In any case, it's easy to create an index from a tree, and that data structure is made for exactly this kind of work. Or in other words, with `index from tree` it's possible to compare against `HEAD` as well.

Reply:
Just pushed an update. I removed `traverse_tree` in favor of the `State` struct built from the head tree. I guess this should be resolved now?

Reply:
Thanks, this gets rid of the manual traversal. 
I'd expect the `gitoxide` status to be faster than the one of `git`  when done, so it's probably better to wait for it instead of implementing everything here.

I wonder how low-hanging the 'avoidance of SHA1 computation' could be similar to how git does it, as I don't know how git efficiently compares the head tree with the index. I hope soon some more analysis will find its way [into the tracking issue](https://github.com/Byron/gitoxide/issues/301).

Reply:
@Byron Thanks for checking! I didn't like the recursion in the first place so it's good to have it removed.

> it's probably better to wait for it instead of implementing everything here

Are you suggesting we wait for `gitoxide` to implement this feature and use it in this PR?

Reply:
Apologies, I guess I was rambling.

I think it's OK to leave this PR as is and look forward to the day when `gitoxide` can perform this task faster than `git` can.